### PR TITLE
Rewrite Hangman & create new struct for loading mode-specific dictionary from sd card

### PR DESCRIPTION
**Pull request by SusanZuo**

### CODE REVIEW

File: SABT_MainUnit/MD4.c [hunk 1]
```diff
@@ -25,293 +36,325 @@
...
+            break;
+            
+        case STATE_GENQUES:
+			
+            num_words = dict.num_words;
+            if (dict.index >= num_words - 1){
+                shuffle(num_words, dict.index_array);
+                dict.index = 0;
+                }
+            chosen_word = dict.words[dict.index_array[dict.index]];
```
Comment (elpintar) [inline SABT_MainUnit/MD4.c]: If in three hint mode and word is three letters, the word is immediately guessed.  Would be nice to add a check here that hints are less than letter count.  Also am curious what happens with double letters (like with "book" and three letters are given, would both "o"s be given too, thus all letters being given?).

File: SABT_MainUnit/MD4.c [hunk 2]
```diff
@@ -25,293 +36,325 @@
...
+            
+        case STATE_LVLSEL:
+			num_mistakes = 0;
+            init_char_arr(mistake_pool, MAX_INCORRECT_GUESS);
+            init_char_arr(input_word, MAX_WORD_LEN);
+            md_last_dot = create_dialog(MP3_LEVEL,
+                                        DOT_1 | DOT_2 | DOT_3 );
+            switch (md_last_dot) {
+					case NO_DOTS:
+						break;
```
Comment (elpintar) [inline SABT_MainUnit/MD4.c]: When trying to press Enter for no hints, the SABT said "invalid pattern".  Maybe some change around here?

File: SABT_MainUnit/dictionary.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ *  Author: Susan Zuo(tzuo)
+ */ 
+
+
+#ifndef DICTIONARY_H_
+#define DICTIONARY_H_
+
+typedef struct dictionary dictionary_t;
+/**
+ *  a struct to store words loaded from SD card
```
Comment (elpintar) [inline SABT_MainUnit/dictionary.h]: Let us know if this is ready to use/how to use, or if it's a work in progress what your plan is.

File: SABT_MainUnit/MD4.c (hunk 1 above)
Comment (SusanZuo) [inline SABT_MainUnit/MD4.c]: Sure, I will add a check in the edge case. Also, in the case mentioned above, only one "o" will be given.

File: SABT_MainUnit/dictionary.h (hunk 3 above)
Comment (SusanZuo) [inline SABT_MainUnit/dictionary.h]: This is still work in progress. Eventually, the teacher should be able to create txt files with specific name on SD card, and the words in the those files will be used as the dictionary for specific mode. Will let you know when it's done

File: SABT_MainUnit/MD4.c (hunk 1 above)
Comment (SusanZuo) [inline SABT_MainUnit/MD4.c]: Check is added. Thx for the feedback.

File: SABT_MainUnit/MD4.c (hunk 2 above)
Comment (SusanZuo) [inline SABT_MainUnit/MD4.c]: Bug fixed now.

File: SABT_MainUnit/dictionary.h (hunk 3 above)
Comment (elpintar) [inline SABT_MainUnit/dictionary.h]: Before you develop too much of this feature - we may want to reprioritize.  Based on the usage of SABTs at the school, I don't think this feature would be super useful to the teachers.  If they want a student to guess a specific word, they usually use two-player hangman.  For spelling many words in a sequence, the teachers have requested a Dictation Mode that would be more useful to them (email me if you need the details of that mode).  Teachers don't often use computers as part of their teaching preparations (or even use computers at all) - so it's unlikely they'll feel comfortable typing up a list of words and putting it on an SD card.  Feel free to reach out about new mode/feature projects - we want to make sure we can make really useful upgrades to the SABTs together the last couple weeks we are here and able to reprogram the SABTs!  Thanks for all the new features you've worked on so far - we're excited to test them with teachers and students soon, and will let you know how it goes.

**Pull request merged**